#213 Split optional deps into extras to shrink dep tree

RobertLD opened the pull request:
Addressing the comment made in #213

jochenchrist commented:
Thanks for your contribution, I agree, this is the way to go.

Futher todos:

- [ ] Not sure about `spark`. This should rather be bound to a server.
- [ ] Add bigquery?
- [ ] Add kafka? (-> requires spark...)
- [ ] Update README.md for an simple "all" installation, like `pip3 install datacontract-cli[bigquery,snowflake,databricks,spark,postgres]
- [ ] Make sure that basic funktionality (lint, diff, import, export?) works without any extra
- [ ] Create a CI/CD test matrix workflow for every extra

RobertLD commented:
> Thanks for your contribution, I agree, this is the way to go.
> 
> Futher todos:
> 
>     * [x]  Not sure about `spark`. This should rather be bound to a server.
> 
>     * [x]  Add bigquery?
> 
>     * [x]  Add kafka? (-> requires spark...)
> 
>     * [ ]  Update README.md for an simple "all" installation, like `pip3 install datacontract-cli[bigquery,snowflake,databricks,spark,postgres]
> 
>     * [ ]  Make sure that basic funktionality (lint, diff, import, export?) works without any extra
> 
>     * [ ]  Create a CI/CD test matrix workflow for every extra

In commit 02119adcc42fb0fd44e33ce6bd44ef9768c3aef1

jochenchrist commented:
I updated the order of the depdencies (reduired extras need to be defined first before being referenced)

RobertLD commented:
> I updated the order of the depdencies (reduired extras need to be defined first before being referenced)


* [x]  Not sure about `spark`. This should rather be bound to a server.

* [x]  Add bigquery?

* [x]  Add kafka? (-> requires spark...)

* [x]  Update README.md for an simple "all" installation, like `pip3 install datacontract-cli[bigquery,snowflake,databricks,spark,postgres]

* [x]  Make sure that basic funktionality (lint, diff, import, export?) works without any extra
** This was tested manually.
* [ ]  Create a CI/CD test matrix workflow for every extra
** Working on this one now

jochenchrist commented:
OK, I played around a bit with pytest markers, but it would significantly increase complexity with just a slight benefit. Let's drop this.

jochenchrist commented:
I updated the Dockerfile and added the Changelog

RobertLD commented:
I had a similar experience trying to configure the CI CD with pytest. The
two ideas I came up with are registering environmental variables that are
matched with pytest skip statements or having the tests be grouped by
folders that match the name of the extra and dynamically running pi test on
those depending on the extras that are installed

On Sat, Jun 1, 2024, 9:00 AM jochenchrist ***@***.***> wrote:

> I updated the Dockerfile and added the Changelog
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/datacontract/datacontract-cli/pull/234#issuecomment-2143440986>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGZJYVXGPPOESOPBCSNYFTZFHAXNAVCNFSM6AAAAABIS32RF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNBTGQ2DAOJYGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

RobertLD commented:
> I updated the Dockerfile and added the Changelog I really appreciate the active contributors on this project 
I'll attach a dependency tree shortly to see what impact we exactly made by implementing the extras

jochenchrist commented:
OK, if you are happy with the reductions, I'd be happy to merge.

RobertLD commented:
> OK, if you are happy with the reductions, I'd be happy to merge.

The core installation now required 570M instead of 1.5Gb which is a pretty vast improvement. 

The largest of which are actual requirements for the core functionality. Perhaps finding a method of removing pyarrow from the core installation would be the final step. I'm happy with merging this as is; as it's still a 3x reduction

jochenchrist commented:
Happy to merge :)
